lte/gateway/c/oai/tasks/s1ap/s1ap_state.cpp

```diff
@@ -155,3 +157,74 @@ void delete_s1ap_ue_state(imsi64_t imsi64) {
   auto imsi_str = S1apStateManager::getInstance().get_imsi_str(imsi64);
   S1apStateManager::getInstance().clear_ue_state_db(imsi_str);
 }
+
+bool get_mme_ue_ids_no_imsi(
+    const hash_key_t keyP, uint64_t const dataP, void* argP, void** resultP) {
+  hash_key_t** mme_id_list   = (hash_key_t**) resultP;
+  uint32_t* num_ues_checked  = (uint32_t*) argP;
+  ue_description_t* ue_ref_p = NULL;
+
+  // Check if a UE reference exists for this comp_s1ap_id
+  hash_table_ts_t* s1ap_ue_state = get_s1ap_ue_state();
+  hashtable_ts_get(s1ap_ue_state, (const hash_key_t) dataP, (void**) &ue_ref_p);
+  if (!ue_ref_p) {
+    *mme_id_list[(*num_ues_checked)++] = keyP;
+    OAILOG_DEBUG(
+        LOG_S1AP,
+        "Adding mme_ue_s1ap_id %lu to eNB clean up list with num_ues_checked "
+        "%u",
+        keyP, *num_ues_checked);
+  }
+}
+
+void remove_ues_without_imsi_from_ue_id_coll() {
+  s1ap_state_t* s1ap_state_p     = get_s1ap_state(false);
+  hashtable_key_array_t* ht_keys = hashtable_ts_get_keys(&s1ap_state_p->enbs);
+  if (ht_keys == nullptr) {
+    return;
+  }
+
+  hashtable_rc_t ht_rc;
+  hash_key_t* mme_ue_id_no_imsi_list;
+  uint32_t num_ues_checked = 0;
+
+  // get each eNB in s1ap_state
+  for (uint32_t i = 0; i < ht_keys->num_keys; i++) {
+    enb_description_t* enb_association_p = nullptr;
+    ht_rc                                = hashtable_ts_get(
+        &s1ap_state_p->enbs, (hash_key_t) ht_keys->keys[i],
+        (void**) &enb_association_p);
+    if (ht_rc != HASH_TABLE_OK) {
+      continue;
+    }
+
+    if (enb_association_p->ue_id_coll.num_elements == 0) {
+      continue;
+    }
+
+    // for each ue comp_s1ap_id in eNB->ue_id_coll, check if it has an S1ap
+    // ue_context, if not delete it
+    mme_ue_id_no_imsi_list = (hash_key_t*) calloc(
+        enb_association_p->ue_id_coll.num_elements, sizeof(hash_key_t));
```

Comment:
shall we define this on the stack itself as an array?

Reply:
@ulaskozat The challenge there is to define the size of this list, as the number of UEs connected to an eNB can vary a lot. What do you think?

Reply:
Let's leave it as is then.